File: ruby/day04_part2.rb
# frozen_string_literal: true

cards = File.read('input/day04.txt').split("\n")
card_copies = Hash.new { |hash, key| hash[key] = 0 }

cards.each.with_index do |card, i|
  numbers = card.split(':').last
  winning_numbers, my_numbers = numbers.split('|').map { |s| s.scan(/\d+/).map(&:to_i) }
  intersection = winning_numbers.intersection(my_numbers).count
  card_copies[i] += 1
  (card_copies[i]).times { ((i + 1)..(i + intersection)).each { |c| card_copies[c] += 1 } }
end

(0...card_copies.count).each { |i| puts "#{i}: #{card_copies[i]}" }

p card_copies.values.reduce(:+)

File: src/bin/day04_part2.rs
use std::{
    collections::{HashMap, HashSet},
    fs,
};

fn main() {
    let input = fs::read_to_string("input/day04.txt").unwrap();
    let cards = input.split('\n');
    let mut card_copies = HashMap::<usize, i32>::new();

    for (card_index, card) in cards.enumerate() {
        let numbers: Vec<Vec<i32>> = card
            .split(':')
            .last()
            .unwrap()
            .split('|')
            .map(|n| {
                n.split_whitespace()
                    .map(|s| s.parse::<i32>().unwrap())
                    .collect()
            })
            .collect();
        let winning_numbers: HashSet<i32> = numbers[0].clone().into_iter().collect();
        let my_numbers: HashSet<i32> = numbers[1].clone().into_iter().collect();
        let intersection = my_numbers.intersection(&winning_numbers).count();

        *card_copies.entry(card_index).or_default() += 1;

        for _ in 0..*card_copies.entry(card_index).or_default() {
            for c in (card_index + 1)..=(card_index + intersection) {
                *card_copies.entry(c).or_default() += 1;
            }
        }
    }

    dbg!(card_copies.values().sum::<i32>());
}
